share/wake/lib/core/json.wake

```diff
@@ -169,3 +169,143 @@ export def x ==/ y = match x y
+
+# Simplify a JSON structure according to the rules given.
+# For example, Wake's implementation of JSON is more lenient than the base
+# standards, so if maximum compatibility is desired in situations where some
+# other implementation may be parsing generated output, `normalizeJSONCompat`
+# may be used to target that lowest common denominator.
+#
+# Note that the relevant normalization is applied to each member of a `JArray`
+# or `JObject` *before* the enclosing container. This means that if some
+# normalization fails anywhere in the JSON tree the entire call will fail, even
+# if the specific branch causing the failure would later have been pruned.
+export def normalizeJSONWith (fmt: JSONNormalize) (json: JValue): Result JValue Error =
+    def normalizeJSON (path: String) = match _
+        JString a =
+            getJSONNormalizeString fmt a
+            | rmap JString
+            | addErrorContext path
+        JInteger a =
+            getJSONNormalizeInteger fmt a
+            | rmap JInteger
+            | addErrorContext path
+        JDouble a =
+            getJSONNormalizeDouble fmt a
+            | rmap JDouble
+            | addErrorContext path
+        JBoolean a =
+            getJSONNormalizeBoolean fmt a
+            | rmap JBoolean
+            | addErrorContext path
+        JNull =
+            Pass JNull
+        JObject a =
+            def normalizeJObjectValue entry =
+                def Pair key value =
+                    entry
+                normalizeJSON "{path}/{key}" value
+                | rmap (Pair key)
+            def normalized =
+                map normalizeJObjectValue a
+            require Pass children =
+                findFail normalized
+            children
+            | getJSONNormalizeObject fmt
+            | rmap JObject
+            | addErrorContext path
+        JArray a =
+            def normalized =
+                zip (seq a.len) a
+                | map (\(Pair i x) normalizeJSON "{path}/[{format i}]" x)
+            require Pass children =
+                findFail normalized
+            children
+            | getJSONNormalizeArray fmt
+            | rmap JArray
+            | addErrorContext path
+    normalizeJSON "" json
+
+# The rules by which `normalizeJSONWith` will simplify JSON values.
+# If any particular function is unable to operate on every input value (e.g. a
+# `JObject` contains duplicate keys of types which cannot be combined), that
+# rule may return a `Fail` which halts the broader processing.
+export tuple JSONNormalize =
+    String: String => Result String Error
+    Integer: Integer => Result Integer Error
+    Double: Double => Result Double Error
+    Boolean: Boolean => Result Boolean Error
+    Object: List (Pair String JValue) => Result (List (Pair String JValue)) Error
+    Array: List JValue => Result (List JValue) Error
+
+# A basic normalization ruleset which returns inputs unchanged.
+export def normalizeJSONIdentity: JSONNormalize =
```

Comment:
> Pretty sure we don't want this one to be the default!
> Maybe call this 'Identity?

Renamed, but I think a noop being the default value of a tuple is legitimate.

Reply:
Sure. I was hoping to have a default already included in `normalizeJSON`, so there was no need to name one of the tuples as `Default`, since the default is whatever `normalizeJSON` passed to `normalizeJSONWith`. I think the right default is what you've called `Compat`.